### ISSUE TITLE
How to retrain the model with new datasets -- segmentation of lesions

### ISSUE THREAD

**Issue by kspruthviraj** [neuropoly/spinalcordtoolbox#2003]
Dear SCT team,
I have installed the latest SC toolbox. And I tried sct_deepseg_lesion on few spinal cord injury dataset (Sagittal, t2 data). And attached an example lesion segmentation

Deep learning--segmented
![dl01](https://user-images.githubusercontent.com/12137274/44233874-0d39c180-a173-11e8-99e1-e4744ac75340.png)

Manual segmentation--
![hr01](https://user-images.githubusercontent.com/12137274/44233875-0d39c180-a173-11e8-8b3f-e92202f89f56.png)

I think the lesion segmentation performs poorly because the model was trained on MS lesion dataset whereas I am testing the model on spinal cord injury (SCI) lesion dataset. 

I have few questions, 

1) If I have to retrain the model, where to find the existing model
or if I build new model, where to find the existing dataset.

2) What kind of labels were used for training the deep learning model ? I mean, manual segmented masks slice-by-slice for the whole volume or only the segmentation from mid-sagittal slice? Because I only have manual segmentations (from SCI dataset) for mid-sagittal slices.

Thank you,
Pruthvi

**Comment by kspruthviraj** [comment]
@charleygros @jcohenadad Few quick questions on retraining the model. 

1) In the paper that you sent **section 2.2, page 8** , could you please help me understand the segmentation framework. Because I have not worked on cascade-based CNN framework so if you have a figure showing the architecture of **CNN2-SC and CNN2-lesion** it would help me to understand better.

2) From the paper, I could see that I would need two ground-truth labels (**segmented lesion and segmented SC**) to segment SC and lesion together. Currently, I only have segmented lesions as ground truth. So,as you have mentioned in your paper, I could use the already existing SCT_deep_seg or Propseg models to segment SC and use it as ground truth. Is that right or am I missing something?

Thank you,
Pruthvi

**Comment by charleygros** [comment]
**Under-segmentation:** a quick tweak of post-processing that you could try is to play with the threshold which aims at binarizing the CNN output:
--> https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L408
You could try 0.001 instead of 0.1: as the decision border of the Dice Loss is quite sharp, you should observe some changes. If you find a satisfying value on your data, please let us know, we could add this parameter in the list of the available parameters of the SCT function.

**Detection out of the SC:** You mentioned that a lesion was detected out of the SC of one case --> very interesting, thank you @kspruthviraj!
I am curious: Would you be able to check if the centerline was correctly detected? (to do this you can re run the segmentation with `-r 0` (i.e. do not remove the temp files) and then check the computed centerline.
If the centerline was correctly detected, we should allow the user to input a SC segmentation, as additional post-processing step.

**Architecture:** _CNN2-SC_ and _CNN2-Lesion_ are trained independently. _CNN2-Lesion_ is used by `sct_deepseg_lesion` while _CNN2-SC_ is used by `sct_deepseg_sc`.
For your application, you will have to retrain _CNN2-Lesion_, using this architecture: `$SCT_DIR/spinalcordtoolbox/deepseg_sc/cnn_models_3d.py`. No need to retrain _CNN1_.
Note: I am realizing that this section is not very clear in the article. I will re-write it and update the ArXiV version by the end of August. Thank you for your feedback @kspruthviraj 

**Re-Training CNN2-Lesion**: Yes it is correct. I would suggest to:
1. Compute the centerline (i.e. center of the spinal cord, 1 voxel per axial slice): 3 options (1) `sct_get_centerline` (2) `sct_deepseg_sc` then `sct_process_segmentation -p centerline` (3) `sct_propseg` then `sct_process_segmentation -p centerline`
2. Set RPI orientation to both the input image and the centerline mask via `sct_image -set-orient RPI`
3. Crop the image around the spinal cord centerline. You can use: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L165
4. Standardize the intensities of the cropped image, via Nyul method, as done here: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L136
5. Set 0.5mm isotropic resolution to both the input image and the centerline mask via `sct_resample -mm 0.5x0.5x0.5` or via https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L520
6. Extract your 3D patches, 48x48x48voxels3
7. Normalize your patch intensities using https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L355. I would recommend to use the mean and std of the initial training dataset, which are for T2: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L382
8. Feed your model! :-)

**Comment by kspruthviraj** [comment]
Regarding detection of SC:
I could not plot the centerline image on the original image due to dimension mismatch. Do you have any function that overcomes this dimension mismatch and plot ?
![screen shot 2018-08-21 at 12 37 21 pm](https://user-images.githubusercontent.com/12137274/44415654-ff06ef00-a53e-11e8-9490-9290b091f9ad.png)

**Comment by jcohenadad** [comment]
@kspruthviraj how did you get the centerline? the output should be in the same space as the input image. Would help if you display the full syntax. thx

**Comment by kspruthviraj** [comment]
@jcohenadad , 

I looked at the temporary folder for the centerline image. There I have only one file named centerline, so I assume this is the one to look at centerline.

**_evans:~ $ /private/var/folders/ft/8cywxd952gx817tbrvpq1rdc0000gs/T/sct-180821121753-9yj41J/bf1_RPI_resampled_centerline_optic.nii**


and then I used this command to display:

fslview /Users/pruthvi_local/Desktop/Lesion_Segmentation/Selected_for_AutoQC_second_copy/bf1.nii -l Greyscale /private/var/folders/ft/8cywxd952gx817tbrvpq1rdc0000gs/T/sct-180821121753-9yj41J/bf1_RPI_resampled_centerline_optic.nii  -l Red -t 0.7

**Comment by jcohenadad** [comment]
@kspruthviraj please do not use the temp image (they are in the temp for good reason ;-) ). Could you pls write the syntax you used to generate the centerline? As @charleygros mentioned you have a few options: https://github.com/neuropoly/spinalcordtoolbox/issues/2003#issuecomment-414557072

**Comment by kspruthviraj** [comment]
@jcohenadad , Oh sorry. Thanks a lot. I missed it. I will check that.

**Comment by kspruthviraj** [comment]
@jcohenadad @charleygros 
This is regarding the cropping of images and extracting 3D patches. As @charleygros mentioned in the previous post, 

**_4. Crop the image around the spinal cord centerline. You can use: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L165
5. Standardize the intensities of the cropped image, via Nyul method, as done here: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L136
6. Extract your 3D patches, 48x48x48voxels3
7. Normalize your patch intensities using https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L355. I would recommend to use the mean and std of the initial training dataset, which are for T2: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L382
Feed your model! :-)_**

Questions:

1) If I crop the original image (i.e. without resampling) around the centerline, then the output I get is okay. However, if I crop the resampled image around the resampled centerline, I get the blank output. And I am bit stuck here. Could you please guide me here.

2) Do you have a function to extract the 3D patches or should I write one ?


Thank you :)

**Comment by jcohenadad** [comment]
@kspruthviraj are you sure the resampling of the centerline did not "null" some voxels (this could happen with nearest neighbor interpolation, because centerline only has a single voxel per slice). In that case better to use trilinear interpolation.

@kspruthviraj by "cropping", do you mean outputting an image with size i,j,k, where i,j are the size of the patch in the i,j directions, and k is the number of slices where the centerline is defined? If so, how is that different from the extraction of 3D patch mentioned in your 2nd point?

**Comment by kspruthviraj** [comment]
@jcohenadad 
1)I checked the resampled output and it has not nulled the voxels and it seems the resampling step worked well. Here are the screenshots,

**Original image:**
![screen shot 2018-09-04 at 3 51 27 pm](https://user-images.githubusercontent.com/12137274/45054273-7b7af100-b05a-11e8-92a3-9fe27b71a0bd.png)

**corresponding centerline image:**
![screen shot 2018-09-04 at 3 21 04 pm](https://user-images.githubusercontent.com/12137274/45052830-3d7bce00-b056-11e8-83e1-d32ae3a3dd37.png)



**Resampled input image:**
![screen shot 2018-09-04 at 3 49 08 pm](https://user-images.githubusercontent.com/12137274/45054155-22ab5880-b05a-11e8-9afd-9c7be6c8d4b2.png)

**Corresponding Resampled centerline:**
![screen shot 2018-09-04 at 3 20 52 pm](https://user-images.githubusercontent.com/12137274/45052846-453b7280-b056-11e8-80cb-5f97b57bd039.png)

2) For feeding the images in to the deep learning model, Charley had suggested to feed in the form of patches, for example (48 x 48 x 48) images. 

In my case, for example, if I have resampled image of size 48 x 48 x 217 or 48 x 48 x 220. For that, I would have to either crop or pad zeros to make different images of same size and then extract the patch with pre-determined size and then feed to the network. This is what I understood from the discussion with @charleygros . May be, I got it wrong. I guess, @charleygros would be able clarify this.

Coming back to first question: From the resampled output it seems that the resampled image looks fine. However, when I try to crop this image with the resampled centerline, I get the blank image. Interestingly, when I use the image that was not resampled and the centerline mask that was not resampled, the cropped image looks fine. Here is the output:

![screen shot 2018-09-04 at 4 03 55 pm](https://user-images.githubusercontent.com/12137274/45054840-2809a280-b05c-11e8-9cbc-823add80a625.png)

**Comment by jcohenadad** [comment]
@kspruthviraj could you please point us to the code that you are using to perform the cropping?

**Comment by kspruthviraj** [comment]
# 5.0 Cropping image

@jcohenadad Here is how I crop:

out_path = '/Users/pruthvi_local/Desktop/Lesion_Segmentation/01_preprocess/04crop/'
PATH_resample = '/Users/pruthvi_local/Desktop/Lesion_Segmentation/01_preprocess/03Resample/'

for line in Mask_list[0:2]:
    Type = line.split("_")
    subject = Type[0]
    print(subject)
    NII_path = PATH_resample + subject + '_resample_INPUT.nii.gz'
    ctr_path = PATH_resample + subject + '_resample_centerline.nii.gz'
    img_NII_path = out_path + subject + '_crop_INPUT.nii'
    **crop_image_around_centerline(NII_path, ctr_path, img_NII_path, 48)**

**Comment by jcohenadad** [comment]
maybe a problem of dtype upon saving? (e.g., binary file saved as uint without appropriate precision). Could you check if the cropped array `data_im_new[:, :, zz]` has non-null voxels? If so, then it is indeed a precision issue upon saving, and you could overcome it by specifying dtype when running `im_new.save(filename_out)`.

**Comment by charleygros** [comment]
@kspruthviraj: As you said, you need to crop the resampled image thanks to the resampled centerline, so that your preprocessing pipeline is robust to data resolution heterogeneity.

I think that your issue comes from this line:
https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_sc.py#L139
--> as your centerline is resampled, it does not contain 1 anymore (but values like 0.99).

Easy solution: Use a nearest neighbor resampling for the centerline: `sct_resample -i t2_centerline -x nn -mm 0.5x0.5x0.5` --> will output a binary resampled centerline
Other solution: Binarize your resampled centerline obtained with linear resampling --> you can use `sct_process_segmentation -p centerline`

**Comment by kspruthviraj** [comment]
@jcohenadad @charleygros Thanks a lot for your answers. I will try with the suggested solutions and will update.

**Comment by charleygros** [comment]
@kspruthviraj : concerning the 3D small patches extraction, you can find some inspiration here: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_deepseg_lesion.py#L160

**Comment by kspruthviraj** [comment]
@charleygros Thank you. The nearest-neighbour trick worked. Now the cropping is running without errors on resampled image.

**Comment by joshuacwnewton** [comment]
Closing as this is not a bug report or feature request. (User-specific use case that looks like it was addressed 2 years ago.)
